Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py

```diff
@@ -1094,7 +1094,45 @@ def checkout(self, argument, prune=None):
                 cwd=self.root_path,
             ).returncode else self.commit()
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
+
+        # Branch is not dev and exists on a remote.
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            local_head = self.commit(branch=argument, include_log=False, include_identifier=False)
+            remote_head = self.commit(branch=path, include_log=False, include_identifier=False)
+            branches_for_remote = self.branches_for(remote_head.hash)
+            local_head_1_ref = '{}~1'.format(argument)
+            local_head_1_hash = run(
+                [self.executable(), 'rev-parse', local_head_1_ref],
+                cwd=self.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            ).stdout.strip()
+
+            # Resets branch if local is not tracking force pushed remote.
+            branches_for_local_1 = self.branches_for(local_head_1_hash)
+            if set(branches_for_local_1) != set(branches_for_remote):
```

Comment:
I think this logic is more complicated than it needs to be. I believe this is as simple as:
```
hsh = run(
    [self.executable(), 'rev-parse', local_head_1_ref],
    cwd=self.root_path,
    capture_output=True,
    encoding='utf-8',
).stdout.strip()[:Commit.HASH_LABEL_SIZE]
bp = self.branch_point(ref= hsh)
if bp.hash != hsh:
    log.info(" You have unsaved changes on the local branch.")
    ...
```

Put another way "Could I make a pull-request from the ref you provided? No? Ok, then you can't possibly have changes you care about associated with it"

Reply:
I was mainly thinking that we don't want to reset the branch unnecessarily if it's already tracking the remote. So by checking the branches associated with the commit, we can see if the remote is tracked. Only if it isn't do we need to check for local changes and reset the branch. Agree with the `branch_point` logic for prompting though. 
```
branches_for_local_1 = self.branches_for(local_head_1_hash)
     if set(branches_for_local_1) != set(branches_for_remote):
          branch_point = self.branch_point(ref=local_head.hash)
          if branch_point != local_head.hash:
               [prompt in terminal]
     [perform normal checkout without reset]
```